// social-api/plugins/twitter.js
/* Misc */
var toolset 		= require('toolset');
var _ 				= require('underscore');
var qs				= require('querystring');
var twitterApi		= require('twitter');

var twitter = function(options, callback) {
	var scope = this;
	
	// Parameters
	this.options 	= _.extend({
		appid:		'',
		appsecret:	''
	},options);
	
	this.url = {
		base:	'https://api.twitter.com/1.1'
	};
	
	this.instance = new twitterApi({
		consumer_key: 			options.consumer_key,
		consumer_secret: 		options.consumer_secret,
		access_token_key: 		options.access_token_key,
		access_token_secret: 	options.access_token_secret
	});
	
	callback(this);
}
twitter.prototype.timeline = function(screen_name, count, callback) {
	this.instance.get('/statuses/user_timeline.json', {
		screen_name:	screen_name,
		count:			count
	}, function(data) {
	   callback(data);
	});
}

twitter.prototype.followers = function(screen_name, count, callback) {
	this.instance.get('/followers/list.json', {
		screen_name:	screen_name,
		count:			count
	}, function(data) {
	   callback(data);
	});
}

module.exports = twitter;
